aiokafka/client.py

```diff
@@ -1,489 +1,348 @@
 
-            # Errors expected for new topics
             try:
-                check_error(topic_metadata)
-            except (UnknownTopicOrPartitionError,
-                    LeaderNotAvailableError) as e:
-
-                # Raise if the topic was passed in explicitly
-                if topic in topics:
-                    raise
-
-                # Otherwise, just log a warning
-                log.error("Error loading topic metadata for %s: %s",
-                          topic, type(e))
+                metadata = yield from conn.send(metadata_request)
+            except KafkaError as err:
+                log.error(
+                    'Unable to request metadata from node with id %s: %s',
+                    node_id, err)
                 continue
 
-            self._topic_partitions[topic] = {}
-            for partition_metadata in partitions:
-                partition = partition_metadata.partition
-                leader = partition_metadata.leader
-
-                self._topic_partitions[topic][partition] = partition_metadata
-
-                # Populate topics_to_brokers dict
-                topic_part = TopicAndPartition(topic, partition)
-
-                # Check for partition errors
-                try:
-                    check_error(partition_metadata)
-
-                # If No Leader, topics_to_brokers topic_partition -> None
-                except LeaderNotAvailableError:
-                    log.error('No leader for topic %s partition %d',
-                              topic, partition)
-                    self._topics_to_brokers[topic_part] = None
-                    continue
-                # If one of the replicas is unavailable -- ignore
-                # this error code is provided for admin purposes only
-                # we never talk to replicas, only the leader
-                except ReplicaNotAvailableError:
-                    log.warning('Some (non-leader) replicas not available '
-                                'for topic %s partition %d', topic, partition)
-                # If Known Broker, topic_partition -> BrokerMetadata
-                if leader in self._brokers:
-                    self._topics_to_brokers[topic_part] = self._brokers[leader]
-
-                # If Unknown Broker, fake BrokerMetadata so we dont lose the id
-                # (not sure how this could happen.
-                # server could be in bad state)
-                else:
-                    self._topics_to_brokers[topic_part] = BrokerMetadata(
-                        leader, None, None
-                    )
+            self.cluster.update_metadata(metadata)
+            break
+        else:
+            log.error('Unable to update metadata from %s', nodeids)
+            self.cluster.failed_update(None)
+            return False
+        return True
 
     @asyncio.coroutine
-    def send_metadata_request(self, payloads=()):
+    def fetch_all_metadata(self):
```

Comment:
this function is almost copy paste of `force_metadata_update`, I think could could be reused.